core-bundle/src/Controller/ContentElement/LoginController.php

```diff
@@ -0,0 +1,186 @@
+use Symfony\Component\HttpFoundation\UriSigner;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+use Symfony\Component\Security\Core\Exception\AuthenticationException;
+use Symfony\Component\Security\Core\Exception\TooManyLoginAttemptsAuthenticationException;
+use Symfony\Component\Security\Http\Authentication\AuthenticationUtils;
+use Symfony\Component\Security\Http\Logout\LogoutUrlGenerator;
+use Symfony\Component\Security\Http\SecurityRequestAttributes;
+use Symfony\Contracts\EventDispatcher\EventDispatcherInterface;
+use Symfony\Contracts\Translation\TranslatorInterface;
+
+#[AsContentElement(category: 'user')]
+class LoginController extends AbstractContentElementController
+{
+    public function __construct(
+        private readonly UriSigner $uriSigner,
+        private readonly LogoutUrlGenerator $logoutUrlGenerator,
+        private readonly AuthenticationUtils $authenticationUtils,
+        private readonly TranslatorInterface $translator,
+        private readonly EventDispatcherInterface $eventDispatcher,
+        private readonly ContaoFramework $framework,
+    ) {
+    }
+
+    protected function getResponse(FragmentTemplate $template, ContentModel $model, Request $request): Response
+    {
+        $user = $this->getUser();
+
+        if ($user && !$user instanceof FrontendUser) {
+            return new Response(status: Response::HTTP_NO_CONTENT);
+        }
+
+        $targetPath = $this->getTargetPath($request);
+        $isRemembered = $this->isGranted('IS_REMEMBERED');
+        $isTwoFactorInProgress = $this->isGranted('IS_AUTHENTICATED_2FA_IN_PROGRESS');
+        $page = $this->getPageModel();
+        $redirect = $request->getUri();
+
+        // The user can re-authenticate on the error_401 page or on the redirect page of
+        // the error_401 page
+        $canReauthenticate = 'error_401' === $page?->type || ($targetPath && $targetPath === $request->query->get('redirect'));
+
+        // Show the logout button if the user is fully authenticated or cannot
+        // re-authenticate on the current page
+        if ($user instanceof FrontendUser && !$isTwoFactorInProgress && (!$isRemembered || !$canReauthenticate)) {
+            // Redirect to last page visited
+            if ($model->redirectBack && $targetPath) {
+                $redirect = $targetPath;
+            }
+
+            // Redirect home if the page is protected
+            elseif ($page?->protected) {
+                $redirect = $request->getSchemeAndHttpHost().$request->getBasePath().'/';
+            }
+
+            $template->logout = true;
+            $template->formId = 'tl_logout_'.$model->id;
+            $template->slabel = $this->translator->trans('MSC.logout', [], 'contao_default');
+            $template->action = $this->logoutUrlGenerator->getLogoutPath();
+
+            // We do not base64_encode the URL here for Symfony's logout controller
+            $template->targetPath = $redirect;
```

Comment:
```suggestion
            if ($redirect) {
                $template->targetPath = $redirect;
            }
```

In the `ModuleLogin` class, `$redirect` is `Environment::get('uri')` in case neither the `if` nor the `elseif` condition matches. But here is can be `null` and I guess it should not be set then?